Make progress bar an optional argument

afermg opened the pull request:
Following up from #67, I moved the tqdm imports inside if-else statements so it is only imported when necessary (in case we would like to make it an optional dependency).

Potential additional changes:
- Propagate the progress_bar argument a few levels up to expose it to the documented API.
- Make tqdm an optional dependency instead of a required one.


Tests run fine:
```
.venv ❯ pytest
================================ test session starts =================================platform linux -- Python 3.11.11, pytest-8.3.4, pluggy-1.5.0
rootdir: /home/amunoz/projects/copairs
configfile: pyproject.toml
plugins: anyio-4.8.0
collected 43 items                                                                   

tests/test_build_rank_multilabel.py .                                          [  2%]
tests/test_compute.py ........                                                 [ 20%]
tests/test_map.py .......                                                      [ 37%]
tests/test_map_filter.py ....                                                  [ 46%]
tests/test_matching.py .........                                               [ 67%]
tests/test_matching_any.py ....                                                [ 76%]
tests/test_matching_multilabel.py ......                                       [ 90%]
tests/test_reference_index.py .                                                [ 93%]
tests/test_replicating.py ...                                                  [100%]

=========================== 43 passed in 73.17s (0:01:13) ============================
```

Some imports were automatically sorted, sorry about that. 

LMK if any more changes are desired.
Alán

afermg commented:
Ready for review @alxndrkalinin

alxndrkalinin commented on src/copairs/compute.py:
```diff
@@ -4,15 +4,16 @@
 
-def parallel_map(par_func: Callable[[int], None], items: np.ndarray) -> None:
+def parallel_map(
+    par_func: Callable[[int], None], items: np.ndarray, progress_bar: bool = True
```
`parallel_map` is used in other places, but the call signature wasn't updated. To make things consistent, this change needs to be propagated upstream, so the progress bar can also be disabled when e.g. `average_precision` is called.

alxndrkalinin reviewed (changes requested):
Thanks for the PR @afermg. I have just one comment regarding access to the new flag from calling `map.average_precision`.

alxndrkalinin commented on tests/test_map.py:
```diff
@@ -114,7 +114,9 @@ def test_pipeline():
     length = len(meta)
     feats = rng.uniform(size=(length, n_feats))
-    average_precision(meta, feats, pos_sameby, pos_diffby, neg_sameby, neg_diffby)
+    average_precision(
```
I think we should also test with False and make sure results match

alxndrkalinin commented on tests/test_map.py:
```diff
@@ -134,7 +136,14 @@ def test_pipeline_multilabel():
     feats = rng.uniform(size=(length, n_feats))
 
     multilabel_average_precision(
```
Same as above

afermg replied:
Added a fixture to test both cases

alxndrkalinin commented on tests/test_map.py:
```diff
@@ -153,9 +162,19 @@ def test_raise_no_pairs():
     feats = rng.uniform(size=(length, n_feats))
     with pytest.raises(UnpairedException, match="Unable to find positive pairs."):
-        average_precision(meta, feats, pos_sameby, pos_diffby, neg_sameby, neg_diffby)
+        average_precision(
```
Same as above

alxndrkalinin reviewed (changes requested):
There few more small issues, and we should updated tests at least for the main 2 endpoints (`average_precision` and `mean_average_precision`), to make sure results are the same as before.

afermg commented on src/copairs/compute.py:
```diff
@@ -4,15 +4,18 @@
+def parallel_map(
+    par_func: Callable[[int], None],
+    items: np.ndarray,
+    progress_bar: Optional[bool] = True,
```
Followed the suggestion and removed it from the type hints.

afermg commented:
I implemented the requested changes. I also added a test that compares the results of with and without progress_bar (pandas' dataframe equality) and it all tests (including the new one and new fixtures) pass

```
.venv ❯ pytest
======================================================== test session starts ========================================================platform linux -- Python 3.11.11, pytest-8.3.4, pluggy-1.5.0
rootdir: /home/amunoz/projects/copairs
configfile: pyproject.toml
plugins: anyio-4.8.0
collected 47 items                                                                                                                  

tests/test_build_rank_multilabel.py .                                                                                         [  2%]
tests/test_compute.py ........                                                                                                [ 19%]
tests/test_map.py ...........                                                                                                 [ 42%]
tests/test_map_filter.py ....                                                                                                 [ 51%]
tests/test_matching.py .........                                                                                              [ 70%]
tests/test_matching_any.py ....                                                                                               [ 78%]
tests/test_matching_multilabel.py ......                                                                                      [ 91%]
tests/test_reference_index.py .                                                                                               [ 93%]
tests/test_replicating.py ...                                                                                                 [100%]

=================================================== 47 passed in 69.71s (0:01:09) 
```

alxndrkalinin commented:
LGTM, I'll let @johnarevalo decide if he wants to take a look
